### PR TITLE
Git fetch should remove the remote tracking refs from working copy if the ref no longer exists on remote. #2194

### PR DESCRIPTION
**Pull request by varshavaradarajan**

### CODE REVIEW

File: common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 1]
```diff
@@ -299,7 +299,7 @@ private String remoteBranch() {
 
     public void fetch(ProcessOutputStreamConsumer outputStreamConsumer) {
         outputStreamConsumer.stdOutput("[GIT] Fetching changes");
-        CommandLine gitFetch = git(environment).withArgs("fetch", "origin").withWorkingDir(workingDir);
+        CommandLine gitFetch = git(environment).withArgs("fetch", "origin", "--prune").withWorkingDir(workingDir);
```
Comment (ketan) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: @varshavaradarajan — we only track one remote branch, why is this required then?

**Comment by varshavaradarajan** [comment]
We do a fit fetch origin. For a non-shallow repo, the refspec is 
```
fetch = +refs/heads/*:refs/remotes/origin/*
```
So even if we have only one branch, ```git fetch origin``` will update all the remote refs kept in the working copy. I thought of changing the refspec while fetching to fetch only the branch specified in the config. But it will break the functionality of submodules.  So, prune will remove the remote ref from the working copy if the remote branch no longer exists.
